akka-http-core/src/main/scala/akka/http/engine/rendering/HttpResponseRendererFactory.scala

```diff
@@ -149,6 +152,11 @@ private[http] class HttpResponseRendererFactory(serverHeader: Option[headers.Ser
 
             if (renderConnectionHeader)
               r ~~ Connection ~~ (if (close) CloseBytes else KeepAliveBytes) ~~ CrLf
+            else if (connHeader != null && connHeader.hasUpgrade && websocketSwitch.isDefined) {
+              r ~~ connHeader ~~ CrLf
+              val websocketHeader = headers.collect { case u: UpgradeToWebsocketResponseHeader ⇒ u }.headOption
+              websocketHeader.foreach(header ⇒ websocketSwitch.get.switchToWebsocket(header.handlerFlow)(header.mat))
```

Comment:
The reason is that it is embedded in the normal routing: you can decide whether you want to support websockets based on all the feature of an Http request, so the natural thing is leaving the decision to the user until the concrete request arrives.